File: SongWei_20170209/piratecat_net/Controls/Chart/Service/DataCenter.cs
/*****************************************************************************\
*                                                                             *
* DataCenter.cs -  Data center functions, types, and definitions.             *
*                                                                             *
*               Version 1.00  ★★★                                          *
*                                                                             *
*               Copyright (c) 2016-2016, Piratecat. All rights reserved.      *
*               Created by Lord 2016/3/10.                                    *
*                                                                             *
*******************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;
using OwLib;
using System.IO;
using System.Net;
using System.Threading;
using System.Xml;
using System.Runtime.InteropServices;
using System.Diagnostics;
using System.Reflection;
using System.Windows;
using System.Windows.Forms;
using OwLibCT;

namespace piratecat
{
    /// <summary>
    /// 处理行情数据
    /// </summary>
    public class DataCenter
    {
        #region Lord 2016/3/10
        /// <summary>
        /// 登录信息
        /// </summary>
        private static LoginInfo m_loginInfo = new LoginInfo();

        /// <summary>
        /// 画线工具
        /// </summary>
        private static Dictionary<String, String> m_plots = new Dictionary<String, String>();

        /// <summary>
        /// 获取画线工具
        /// </summary>
        public static Dictionary<String, String> Plots
        {
            get { return m_plots; }
        }

        private static ChatServiceEx m_chatService = new ChatServiceEx();

        /// <summary>
        /// 获取或设置聊天服务
        /// </summary>
        public static ChatServiceEx ChatService
        {
            get { return m_chatService; }
        }

        private static ConnectService m_connectService = new ConnectService();

        /// <summary>
        /// 获取连接服务
        /// </summary>
        public static ConnectService ConnectService
        {
            get { return DataCenter.m_connectService; }
        }

        private static int m_connectState = 0;

        /// <summary>
        /// 获取连接状态
        /// </summary>
        public static int ConnectState
        {
            get { return DataCenter.m_connectState; }
        }

        private static IndicatorServiceEx m_indicatorService = new IndicatorServiceEx();

        /// <summary>
        /// 获取或设置指标服务
        /// </summary>
        public static IndicatorServiceEx IndicatorService
        {
            get { return DataCenter.m_indicatorService; }
        }

        private static IndicatorLayoutServiceEx m_indicatorLayoutService = new IndicatorLayoutServiceEx();

        /// <summary>
        /// 获取或设置指标布局服务
        /// </summary>
        public static IndicatorLayoutServiceEx IndicatorLayoutService
        {
            get { return DataCenter.m_indicatorLayoutService; }
        }

        private static LoginService m_loginService = new LoginService();

        /// <summary>
        /// 获取登录服务
        /// </summary>
        public static LoginService LoginService
        {
            get { return DataCenter.m_loginService; }
        }

        private static MacroServiceEx m_macroService = new MacroServiceEx();

        /// <summary>
        /// 获取宏服务
        /// </summary>
        public static MacroServiceEx MacroService
        {
            get { return DataCenter.m_macroService; }
        }

        private static UIXmlEx m_mainUI;

        /// <summary>
        /// 设置主控件
        /// </summary>
        public static UIXmlEx MainUI
        {
            set { DataCenter.m_mainUI = value; }
        }

        private static ProxyService m_proxyService = new ProxyService();

        /// <summary>
        /// 获取代理连接信息
        /// </summary>
        public static ProxyService ProxyService
        {
            get { return DataCenter.m_proxyService; }
        }

        private static QuoteService m_quoteService = new QuoteService();

        /// <summary>
        /// 获取行情服务
        /// </summary>
        public static QuoteService QuoteService
        {
            get { return DataCenter.m_quoteService; }
        }

        private static SecurityFilterServiceEx m_securityFilterService = new SecurityFilterServiceEx();

        /// <summary>
        /// 获取条件选股服务
        /// </summary>
        public static SecurityFilterServiceEx SecurityFilterService
        {
            get { return DataCenter.m_securityFilterService; }
        }

        private static SecurityServiceEx m_securityService = new SecurityServiceEx();

        /// <summary>
        /// 获取证券服务
        /// </summary>
        public static SecurityServiceEx SecurityService
        {
            get { return DataCenter.m_securityService; }
        }

        private static ServerService m_serverService = new ServerService();

        /// <summary>
        /// 获取服务器服务
        /// </summary>
        public static ServerService ServerService
        {
            get { return DataCenter.m_serverService; }
        }

        private static int m_userID = -1;

        /// <summary>
        /// 获取用户ID
        /// </summary>
        public static int UserID
        {
            get { return m_userID; }
        }

        private static UserCookieService m_userCookieService = new UserCookieService();

        /// <summary>
        /// 用户Cookie服务
        /// </summary>
        public static UserCookieService UserCookieService
        {
            get { return DataCenter.m_userCookieService; }
        }

        private static UserSecurityServiceEx m_userSecurityService = new UserSecurityServiceEx();

        /// <summary>
        /// 获取或设置自选股服务
        /// </summary>
        public static UserSecurityServiceEx UserSecurityService
        {
            get { return DataCenter.m_userSecurityService; }
        }

        private static UserSessionServiceEx m_userSessionService = new UserSessionServiceEx();

        /// <summary>
        /// 获取用户会话服务
        /// </summary>
        public static UserSessionServiceEx UserSessionService
        {
            get { return DataCenter.m_userSessionService; }
        }

        /// <summary>
        /// 连接到服务器
        /// </summary>
        /// <returns>状态</returns>
        public static int Connect()
        {
            int socketID = -1;
            int qtSocketID = -1;
            List<ConnectInfo> connectInfos = new List<ConnectInfo>();
            m_connectService.GetConnects(connectInfos);
            int connectInfosSize = connectInfos.Count;
            for (int i = 0; i < connectInfosSize; i++)
            {
                ConnectInfo connectInfo = connectInfos[i];
                if (connectInfo.m_type == "主服务器")
                {
                    List<ProxyInfo> proxyInfos = new List<ProxyInfo>();
                    m_proxyService.GetProxyInfos(proxyInfos);
                    int proxyInfosSize = proxyInfos.Count;
                    for (int j = 0; j < proxyInfosSize; j++)
                    {
                        ProxyInfo pInfo = proxyInfos[i];
                        if (pInfo.m_use)
                        {
                            String ip = connectInfo.m_ip, proxyIP = "", proxyUserName = pInfo.m_name, proxyPwd = pInfo.m_pwd, proxyDomain = pInfo.m_domain;
					         int proxyPort = pInfo.m_port;
					        if(pInfo.m_type == 3)
					        {
						        //CurlHttp *http = CurlHttp::GetInstance();
						        //http->GetIeProxySet(&proxyIP, &proxyPort);
					        }
					        else
					        {
						        proxyIP = pInfo.m_ip;
					        }
					        int cType = 0;
					        if(pInfo.m_type == 1 || pInfo.m_type == 3)
					        {
						        cType = 1;
					        }
					        else if(pInfo.m_type == 2)
					        {
						        cType = 3;
					        }
                            socketID = BaseService.ConnectToServer(cType, ip, connectInfo.m_port, proxyIP, proxyPort, proxyUserName, proxyPwd, proxyDomain, 6);
                            qtSocketID = BaseService.ConnectToServer(cType, ip, connectInfo.m_port, proxyIP, proxyPort, proxyUserName, proxyPwd, proxyDomain, 6);
                            break;
                        }
                    }                  
                }
            }
            if (socketID > 0)
            {
                m_connectState = 1;
                m_loginService.SocketID = socketID;
                m_indicatorService.SocketID = socketID;
                m_quoteService.SocketID = qtSocketID;
                m_securityService.SocketID = socketID;
                m_chatService.SocketID = socketID;
                m_indicatorLayoutService.SocketID = socketID;
                m_macroService.SocketID = socketID;
                m_securityFilterService.SocketID = socketID;
                m_userSecurityService.SocketID = socketID;
                m_userSessionService.SocketID = socketID;
            }
            return socketID;
        }

        /// <summary>
        /// 取消连接
        /// </summary>
        /// <returns>状态</returns>
        public static int DisConnect()
        {
            m_connectState = 0;
            BaseService.CloseClient(m_loginService.SocketID);
            BaseService.CloseClient(m_quoteService.SocketID);
            return 1;
        }

        /// <summary>
        /// 获取程序路径
        /// </summary>
        /// <returns>程序路径</returns>
        public static String GetAppPath()
        {
            return Application.StartupPath;
        }

        /// <summary>
        /// 获取用户目录
        /// </summary>
        /// <returns>用户目录</returns>
        public static String GetUserPath()
        {
            String userPath = Environment.GetEnvironmentVariable("LOCALAPPDATA");
	        if(!CFileA.IsDirectoryExist(userPath))
	        {
		        userPath = GetAppPath();
	        }
	        else
	        {
		        userPath += "\\piratecat2";
		        if(!CFileA.IsDirectoryExist(userPath))
		        {
			        CFileA.CreateDirectory(userPath);
		        }
	        }
	        return userPath;
            
        }

        /// <summary>
        /// 加载数据
        /// </summary>
        /// <param name="state">状态</param>
        /// <returns>加载状态</returns>
        public static int LoadData(int state)
        {
            ChatData chatData = new ChatData();
            m_chatService.Enter(m_chatService.RequestID, chatData);
            Indicator indicator = new Indicator();
            indicator.m_userID = m_userID;
            m_indicatorService.GetIndicators(m_indicatorService.GetListRequestID, indicator);
            SecurityFilterTemplate template = new SecurityFilterTemplate();
            template.m_userID = m_userID;
            m_securityFilterService.GetTemplates(m_securityFilterService.GetListRequestID, template);
            UserSecurityVisitsCount visitsCount = new UserSecurityVisitsCount();
            visitsCount.m_userID = DataCenter.UserID;
            m_userSecurityService.GetVisitsCount(m_userSecurityService.GetListRequestID, visitsCount);
            UserSecurityCategory category = new UserSecurityCategory();
            category.m_userID = m_userID;
            m_userSecurityService.GetCategories(m_userSecurityService.GetListRequestID, category);
            IndicatorLayout indicatorLayout = new IndicatorLayout();
            indicatorLayout.m_userID = m_userID;
            m_indicatorLayoutService.GetLayouts(m_indicatorLayoutService.GetListRequestID, indicatorLayout);
            Macro macro = new Macro();
            macro.m_userID = m_userID;
            m_macroService.GetMacros(m_macroService.GetListRequestID, macro);
            UserSession userSession = new UserSession();
            userSession.m_userID = m_userID;
            m_userSessionService.GetSessions(m_userSessionService.GetListRequestID, userSession);
            if (state == 1)
            {
                m_securityService.LoadSecurities();
            }
            if (m_mainUI != null)
            {
                m_mainUI.LoadData();
            }
            return 0;
        }

        /// <summary>
        /// 重连方法
        /// </summary>
        private static void OnReconnect()
        {
            long downFLow = 0;
            while(true)
            {
                long newDownFLow = BaseService.DownFlow;
                if(downFLow > 0)
                {
                    if(downFLow == newDownFLow)
                    {
                        DisConnect();
                        while (true)
                        {
                            if(Connect() > 0)
                            {
                                int loginOutReqID = BaseService.GetRequestID();
                                int loginReqID = BaseService.GetRequestID();
                                m_loginService.Logout(loginOutReqID);
                                CMessage message = new CMessage();
                                m_loginService.RegisterWait(loginReqID, message);
                                m_loginService.Login(m_loginInfo.m_userName, m_loginInfo.m_passWord, loginReqID);
                                if(m_loginService.WaitMessage(loginReqID, 10000) > 0)
                                {
                                    List<LoginInfo> loginInfos = new List<LoginInfo>();
                                    LoginService.GetLoginInfos(loginInfos, message.m_body, message.m_bodyLength);
                                    int loginInfosSize = loginInfos.Count;
                                    if(loginInfosSize > 0)
                                    {
                                        LoginInfo loginInfo = loginInfos[0];
                                        if(loginInfo.m_state == 1)
                                        {
                                            SaveLoginInfo(loginInfo);
                                            LoadData(2);
                                        }
                                    }
                                    loginInfos.Clear();
                                }
                                else
                                {
                                    m_loginService.Logout(loginOutReqID);
                                }
                                break;
                            }
                            Thread.Sleep(1000);
                        }
                    }
                }
                downFLow = newDownFLow;
                Thread.Sleep(10000);
            }
        }

        /// <summary>
        /// 读取所有的画线工具
        /// </summary>
        private static void ReadPlots()
        {
            String xmlPath = Path.Combine(GetAppPath(), "config\\Plots.xml");
            m_plots.Clear();
            if (File.Exists(xmlPath))
            {
                XmlDocument xmlDoc = new XmlDocument();
                xmlDoc.Load(xmlPath);
                XmlNode rootNode = xmlDoc.DocumentElement;
                foreach (XmlNode node in rootNode.ChildNodes)
                {
                    if (node.Name.ToUpper() == "PLOT")
                    {
                        String name = String.Empty;
                        String text = String.Empty;
                        foreach (XmlNode childeNode in node.ChildNodes)
                        {
                            if (childeNode.Name.ToUpper() == "NAME")
                            {
                                name = childeNode.InnerText;
                            }
                            else if (childeNode.Name.ToUpper() == "TEXT")
                            {
                                text = childeNode.InnerText;
                            }
                        }
                        m_plots[name] = text;
                    }
                }
            }
        }

        /// <summary>
        /// 保存登录信息
        /// </summary>
        /// <param name="sessionID">会话ID</param>
        public static void SaveLoginInfo(LoginInfo loginInfo)
        {
            if (m_userID == -1)
            {
                Thread hThread = new Thread(new ThreadStart(OnReconnect));
                hThread.IsBackground = true;
                hThread.Start();
            }
            m_loginInfo = loginInfo;
            m_userID = loginInfo.m_userID;
            m_chatService.SessionID = loginInfo.m_sessionID;
            m_indicatorLayoutService.SessionID = loginInfo.m_sessionID;
            m_indicatorService.SessionID = loginInfo.m_sessionID;
            m_loginService.SessionID = loginInfo.m_sessionID;
            m_macroService.SessionID = loginInfo.m_sessionID;
            m_quoteService.SessionID = loginInfo.m_sessionID;
            m_securityFilterService.SessionID = loginInfo.m_sessionID;
            m_securityService.SessionID = loginInfo.m_sessionID;
            m_userCookieService.UserID = m_userID;
            m_userSecurityService.SessionID = loginInfo.m_sessionID;
            m_userSessionService.SessionID = loginInfo.m_sessionID;
        }

        /// <summary>
        /// 开启服务
        /// </summary>
        /// <param name="appPath">程序路径</param>
        public static void StartService()
        {
            BaseService.AddService(m_chatService);
            BaseService.AddService(m_indicatorLayoutService);
            BaseService.AddService(m_indicatorService);
            BaseService.AddService(m_loginService);
            BaseService.AddService(m_macroService);
            BaseService.AddService(m_quoteService);
            BaseService.AddService(m_securityFilterService);
            BaseService.AddService(m_securityService);
            BaseService.AddService(m_userSecurityService);
            BaseService.AddService(m_userSessionService);
            //List<BaseService> services = new List<BaseService>();
            //BaseService.GetServices(services);
            //int servicesSize = services.Count;
            //for (int i = 0; i < servicesSize; i++)
            //{
            //    BaseService service = services[i];
            //    service.CompressType = BaseService.COMPRESSTYPE_NONE;
            //}
            ReadPlots();
        }
        #endregion
    }
}

File: SongWei_20170209/piratecat/piratecat/Chart/Service/DataCenter.cpp
#include "..\\..\\stdafx.h"
#include "DataCenter.h"

static AIService *m_aiService = 0;
static ChatServiceEx *m_chatService = 0;
static ConnectService *m_connectService = 0;
static int m_connectState = 0;
static IndicatorLayoutServiceEx *m_indicatorLayoutService = 0;
static IndicatorServiceEx *m_indicatorService = 0;
static LoginInfo m_loginInfo;
static LoginService *m_loginService = 0;
static MacroServiceEx *m_macroService = 0;
static NewsService *m_newsService = 0;
static UIXmlEx *m_mainUI = 0;
static ProxyService *m_proxyService = 0;
static QuoteService *m_quoteService = 0;
static SecurityServiceEx *m_securityService = 0;
static SecurityFilterServiceEx *m_securityFilterService = 0;
static ServerService *m_serverService = 0;
static int m_userID = -1;
static UserCookieService* m_userCookieService = 0;
static UserSecurityServiceEx *m_userSecurityService = 0;
static UserSessionServiceEx *m_userSessionService = 0;
static map<String, Security> m_securities;

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

DataCenter::DataCenter()
{
}

DataCenter::~DataCenter()
{
}

void DataCenter::dprintf( char *format, ...)
{
	static char buf[1024];
	va_list args;
	va_start( args, format );
	vsprintf_s( buf, format, args );
	va_end( args);
	OutputDebugStringA( buf);
	printf("%s", buf);
}

IXMLDOMDocument* DataCenter::DomFromCOM()
{
   HRESULT hr;
   IXMLDOMDocument *pxmldoc = 0;
 
   HRCALL( CoCreateInstance(__uuidof(MSXML2::DOMDocument30),
				  NULL,
				  CLSCTX_INPROC_SERVER,
				  __uuidof(IXMLDOMDocument),
				  (void**)&pxmldoc),
				  "Create a new DOMDocument");
 
	HRCALL( pxmldoc->put_async(VARIANT_FALSE),
			"should never fail");
	HRCALL( pxmldoc->put_validateOnParse(VARIANT_FALSE),
			"should never fail");
	HRCALL( pxmldoc->put_resolveExternals(VARIANT_FALSE),
			"should never fail");
 
	return pxmldoc;
clean:
	if (pxmldoc)
	{
		pxmldoc->Release();
	}
	return 0;
}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

AIService*  DataCenter::GetAIService()
{
	return m_aiService;
}

string DataCenter::GetAppPath()
{   
    char exeFullPath[MAX_PATH]; 
    string strPath = "";
    GetModuleFileNameA(0, exeFullPath, MAX_PATH);
    strPath = (string)exeFullPath; 
    int pos = (int)strPath.find_last_of('\\', strPath.length());
    return strPath.substr(0, pos);
}

ConnectService* DataCenter::GetConnectService()
{
	return m_connectService;
}

int DataCenter::GetConnectState()
{
	return m_connectState;
}

string DataCenter::GetUserPath()
{	
	char buffer[1024] = {0};
	::GetEnvironmentVariableA("LOCALAPPDATA", buffer, 1024);
	string userPath = buffer;
	if(!CFileA::IsDirectoryExist(userPath.c_str()))
	{
		userPath = GetAppPath();
	}
	else
	{
		userPath += "\\piratecat";
		if(!CFileA::IsDirectoryExist(userPath.c_str()))
		{
			CFileA::CreateDirectory(userPath.c_str());
		}
	}
	return userPath;
}

map<String, String> DataCenter::GetPlots()
{
	map<String, String> plots;
	IXMLDOMDocument *pXMLDom = 0;
    IXMLDOMParseError *pXMLErr = 0;
	IXMLDOMElement *spElement = 0;
	IXMLDOMNodeList *spNodeList = 0;
    BSTR bstr = 0;
    VARIANT_BOOL status;
    VARIANT var;
    HRESULT hr;
	String wPath;
	string path = GetAppPath() + "\\config\\Plots.xml";
	CStr::stringTowstring(wPath, path);
	long lCount = 0;
    pXMLDom = DomFromCOM();
    if (!pXMLDom) goto clean;
	VariantInit(&var);
	V_BSTR(&var) = SysAllocString(wPath.c_str());
	V_VT(&var) = VT_BSTR;
	HRCALL(pXMLDom->load(var, &status), "");
	if (status != VARIANT_TRUE) {
		HRCALL(pXMLDom->get_parseError(&pXMLErr),"");
		HRCALL(pXMLErr->get_reason(&bstr),"");
		goto clean;
	}
	HRCALL(pXMLDom->get_xml(&bstr), "");
	hr = pXMLDom->get_documentElement(&spElement);
	hr = spElement->get_childNodes(&spNodeList); 
	hr = spNodeList->get_length(&lCount);  
	for (long i = 0; i < lCount; ++i)
	{
		IXMLDOMNode *spChildNode = 0;
		hr = spNodeList->get_item(i, &spChildNode); 
		BSTR name;
		hr = spChildNode->get_nodeName(&name);
		String strName = name;
		if(name)
		{
			SysFreeString(name);
		}
		if(strName == L"Plot")
		{
			IXMLDOMNodeList *spChildNodeList = 0;
			hr = spChildNode->get_childNodes(&spChildNodeList);
			long childCount = 0;
			hr = spChildNodeList->get_length(&childCount);
			String plotName;
			String plotText;
			for(long j = 0; j < childCount; j++)
			{
				IXMLDOMNode *spChildNode2 = 0;
				hr = spChildNodeList->get_item(j, &spChildNode2);
				BSTR name2;
				hr = spChildNode2->get_nodeName(&name2);
				String strName2 = name2;
				if(name2)
				{
					SysFreeString(name2);
				}
				BSTR value;
				spChildNode2->get_text(&value);
				String strValue = value;
				if(value)
				{
					SysFreeString(value);
				}
				if(strName2 == L"Name")
				{
					plotName = strValue;
				}
				else if(strName2 == L"Text")
				{
					plotText = strValue;
				}
				if(spChildNode2)
				{
					spChildNode2->Release();
				}
			}
			plots[plotName] = plotText;
			if(spChildNodeList)
			{
				spChildNodeList->Release();
			}
		}
		if(spChildNode)
		{
			spChildNode->Release();
		}
	}
	clean:
    if (bstr) SysFreeString(bstr);
    if (&var) VariantClear(&var);
    if (pXMLErr) pXMLErr->Release();
    if (pXMLDom) pXMLDom->Release();
	if (spElement) spElement->Release();
	if (spNodeList) spNodeList->Release();
	return plots;
}

ChatServiceEx* DataCenter::GetChatService()
{
	return m_chatService;
}

IndicatorLayoutServiceEx* DataCenter::GetIndicatorLayoutService()
{
	return m_indicatorLayoutService;
}

IndicatorServiceEx* DataCenter::GetIndicatorService()
{
	return m_indicatorService;
}

LoginService* DataCenter::GetLoginService()
{
	return m_loginService;
}

MacroServiceEx* DataCenter::GetMacroService()
{
	return m_macroService;
}

NewsService* DataCenter::GetNewsService()
{
	return m_newsService;
}

ProxyService* DataCenter::GetProxyService()
{
	return m_proxyService;
}

QuoteService* DataCenter::GetQuoteService()
{
	return m_quoteService;
}

SecurityFilterServiceEx* DataCenter::GetSecurityFilterService()
{
	return m_securityFilterService;
}

SecurityServiceEx* DataCenter::GetSecurityService()
{
	return m_securityService;
}

ServerService* DataCenter::GetServerService()
{
	return m_serverService;
}

int DataCenter::GetUserID()
{
	return m_userID;
}

UserCookieService* DataCenter::GetUserCookieService()
{
	return m_userCookieService;
}

UserSecurityServiceEx* DataCenter::GetUserSecurityService()
{
	return m_userSecurityService;
}

UserSessionServiceEx* DataCenter::GetUserSessionService()
{
	return m_userSessionService;
}

void DataCenter::SetMainUI(UIXmlEx *mainUI)
{
	m_mainUI = mainUI;
}

/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

int DataCenter::Connect()
{
	int socketID = -1;
	int qtSocketID = -1;
	vector<ConnectInfo> connectInfos;
	m_connectService->GetConnects(&connectInfos);
	vector<ConnectInfo>::iterator sIter = connectInfos.begin();
	for(; sIter != connectInfos.end(); ++sIter)
	{
		ConnectInfo connectInfo = *sIter;
		if(connectInfo.m_type == L"Ö÷·þÎñÆ÷")
		{
			vector<ProxyInfo> proxyInfos;
			m_proxyService->GetProxyInfos(&proxyInfos);
			vector<ProxyInfo>::iterator sIter2 = proxyInfos.begin();
			for(; sIter2 != proxyInfos.end(); ++sIter2)
			{
				ProxyInfo pInfo = *sIter2;
				if(pInfo.m_use)
				{
					string ip = "", proxyIP = "", proxyUserName = "", proxyPwd = "", proxyDomain = "";
					CStrA::wstringTostring(ip, connectInfo.m_ip);
					CStrA::wstringTostring(proxyUserName, pInfo.m_name);
					CStrA::wstringTostring(proxyPwd, pInfo.m_pwd);
					CStrA::wstringTostring(proxyDomain, pInfo.m_domain);
					 int proxyPort = pInfo.m_port;
					if(pInfo.m_type == 3)
					{
						//CurlHttp *http = CurlHttp::GetInstance();
						//http->GetIeProxySet(&proxyIP, &proxyPort);
					}
					else
					{
						CStrA::wstringTostring(proxyIP, pInfo.m_ip);
					}
					int cType = 0;
					if(pInfo.m_type == 1 || pInfo.m_type == 3)
					{
						cType = 1;
					}
					else if(pInfo.m_type == 2)
					{
						cType = 3;
					}
					socketID = BaseService::ConnectToServer(cType, ip.c_str(), connectInfo.m_port, proxyIP.c_str(), proxyPort, proxyUserName.c_str(), proxyPwd.c_str(), proxyDomain.c_str(), 6);
					qtSocketID = BaseService::ConnectToServer(cType, ip.c_str(), connectInfo.m_port, proxyIP.c_str(), proxyPort, proxyUserName.c_str(), proxyPwd.c_str(), proxyDomain.c_str(), 6);
					break;
				}
			}
		}
	}
	if(socketID > 0)
	{
		m_connectState = 1;
		m_aiService->SetSocketID(socketID);
		m_loginService->SetSocketID(socketID);
		m_newsService->SetSocketID(socketID);
		m_quoteService->SetSocketID(qtSocketID);
		m_indicatorService->SetSocketID(qtSocketID);
		m_securityService->SetSocketID(socketID);
		m_chatService->SetSocketID(socketID);
		m_indicatorLayoutService->SetSocketID(socketID);
		m_macroService->SetSocketID(socketID);
		m_securityFilterService->SetSocketID(socketID);
		m_userSecurityService->SetSocketID(socketID);
		m_userSessionService->SetSocketID(socketID);
	}
	return socketID;
}

int DataCenter::DisConnect()
{
	m_connectState = 0;
	BaseService::CloseClient(m_loginService->GetSocketID());
	BaseService::CloseClient(m_quoteService->GetSocketID());
	return 1;
}

int DataCenter::LoadData(int state)
{
	ChatData chatData;
	m_chatService->Enter(m_chatService->GetRequestID(), &chatData);
	Indicator indicator;
	indicator.m_userID = m_userID;
	(dynamic_cast<IndicatorService*>(m_indicatorService))->GetIndicators(m_indicatorService->GetListRequestID(), &indicator);
	SecurityFilterTemplate sfTemplate;
	sfTemplate.m_userID = m_userID;
	(dynamic_cast<SecurityFilterService*>(m_securityFilterService))->GetTemplates(m_securityFilterService->GetListRequestID(), &sfTemplate);
	UserSecurityVisitsCount visitsCount;
    visitsCount.m_userID = m_userID;
    m_userSecurityService->GetVisitsCount(m_userSecurityService->GetListRequestID(), &visitsCount);
	Macro macro;
	macro.m_userID = m_userID;
	(dynamic_cast<MacroService*>(m_macroService))->GetMacros(m_macroService->GetListRequestID(), &macro);
	UserSecurityCategory category;
	category.m_userID = m_userID;
	(dynamic_cast<UserSecurityService*>(m_userSecurityService))->GetCategories(m_userSecurityService->GetListRequestID(), &category);
	IndicatorLayout indicatorLayout;
	indicatorLayout.m_userID = m_userID;
	(dynamic_cast<IndicatorLayoutService*>(m_indicatorLayoutService))->GetLayouts(m_indicatorLayoutService->GetListRequestID(), &indicatorLayout);
	UserSession userSession;
    userSession.m_userID = m_userID;
    (dynamic_cast<UserSessionService*>(m_userSessionService))->GetSessions(m_userSessionService->GetListRequestID(), &userSession);
	if(state == 1)
	{
		m_securityService->LoadSecurities();
	}
	if(m_mainUI)
	{
		m_mainUI->LoadData();
	}
	return 0;
}

DWORD WINAPI OnReconnect(LPVOID lpParam)
{
	_int64 downFLow = 0;
    while(1)
    {
        _int64 newDownFLow = BaseService::GetDownFlow();
        if(downFLow > 0)
        {
            if(downFLow == newDownFLow)
            {
                DataCenter::DisConnect();
                while (1)
                {
                    if(DataCenter::Connect() > 0)
                    {
                        int loginOutReqID = BaseService::GetRequestID();
                        int loginReqID = BaseService::GetRequestID();
                        m_loginService->Logout(loginOutReqID);
                        CMessage message;
                        m_loginService->RegisterWait(loginReqID, &message);
                        m_loginService->Login(m_loginInfo.m_userName, m_loginInfo.m_passWord, loginReqID);
                        if(m_loginService->WaitMessage(loginReqID, 10000) > 0)
                        {
                            vector<LoginInfo> loginInfos;
                            LoginService::GetLoginInfos(&loginInfos, message.m_body, message.m_bodyLength);
                            int loginInfosSize = (int)loginInfos.size();
                            if(loginInfosSize > 0)
                            {
                                LoginInfo loginInfo = loginInfos[0];
                                if(loginInfo.m_state == 1)
                                {
                                    DataCenter::SaveLoginInfo(&loginInfo);
                                    DataCenter::LoadData(2);
                                }
                            }
                            loginInfos.clear();
                        }
                        else
                        {
                            m_loginService->Logout(loginOutReqID);
                        }
                        break;
                    }
                    Sleep(1000);
                }
            }
        }
        downFLow = newDownFLow;
        Sleep(10000);
    }
    return 1;
}

void DataCenter::SaveLoginInfo(LoginInfo* loginInfo)
{
	if(m_userID == -1)
	{
		HANDLE hThread = ::CreateThread(0, 0, OnReconnect, 0, 0, 0);
		::CloseHandle(hThread);
	}
	m_loginInfo = *loginInfo;
    m_userID = loginInfo->m_userID;
	m_aiService->SetSessionID(loginInfo->m_sessionID);
	m_chatService->SetSessionID(loginInfo->m_sessionID);
	m_indicatorLayoutService->SetSessionID(loginInfo->m_sessionID);
    m_indicatorService->SetSessionID(loginInfo->m_sessionID);
	m_loginService->SetSessionID(loginInfo->m_sessionID);
	m_macroService->SetSessionID(loginInfo->m_sessionID);
	m_newsService->SetSessionID(loginInfo->m_sessionID);
    m_quoteService->SetSessionID(loginInfo->m_sessionID);
	m_securityFilterService->SetSessionID(loginInfo->m_sessionID);
    m_securityService->SetSessionID(loginInfo->m_sessionID);
	m_userCookieService->SetUserID(m_userID);
    m_userSecurityService->SetSessionID(loginInfo->m_sessionID);
	m_userSessionService->SetSessionID(loginInfo->m_sessionID);
}

void DataCenter::StartService()
{
	m_aiService = new AIService;
	m_chatService = new ChatServiceEx;
	m_connectService = new ConnectService;
	m_indicatorLayoutService = new IndicatorLayoutServiceEx;
	m_indicatorService = new IndicatorServiceEx;
	m_loginService = new LoginService;
	m_macroService = new MacroServiceEx;
	m_newsService = new NewsService;
	m_proxyService = new ProxyService;
	m_quoteService = new QuoteService;
	m_securityService = new SecurityServiceEx;
	m_securityFilterService = new SecurityFilterServiceEx;
	m_serverService = new ServerService;
	m_userCookieService = new UserCookieService;
	m_userSecurityService = new UserSecurityServiceEx;
	m_userSessionService = new UserSessionServiceEx;
	BaseService::AddService(m_aiService);
	BaseService::AddService(m_chatService);
	BaseService::AddService(m_indicatorLayoutService);
	BaseService::AddService(m_indicatorService);
	BaseService::AddService(m_loginService);
	BaseService::AddService(m_macroService);
	BaseService::AddService(m_newsService);
	BaseService::AddService(m_quoteService);
	BaseService::AddService(m_securityFilterService);
	BaseService::AddService(m_securityService);
	BaseService::AddService(m_userSecurityService);
	BaseService::AddService(m_userSessionService);
	vector<BaseService*> services;
	BaseService::GetServices(&services);
	vector<BaseService*>::iterator sIter = services.begin();
	for(; sIter != services.end(); ++sIter)
	{
		BaseService *service = *sIter;
		service->SetCompressType(COMPRESSTYPE_NONE);
	}
}

